Handle tokens without price in rounding buffer calculation

e00E opened the pull request:
The rounding buffer calculation takes token prices as input. If we don't
have a price for an order we need to pick some default or we can't
perform the calculation.
Previously we used `1` as the default but it seems more correct to reuse
the fee price (10e18) instead of 1. Either way is a guess but this is a
better heuristic.

### Test Plan
This function is not yet used.

e00E committed:
Handle tokens without price in rounding buffer calculation

The rounding buffer calculation takes token prices as input. If we don't
have a price for an order we need to pick some default or we can't
perform the calculation.
Previously we used `1` as the default but it seems more correct to reuse
the fee price (10e18) instead of 1. Either way is a guess but this is a
better heuristic.

fleupold commented on price-estimator/src/solver_rounding_buffer.rs:
```diff
@@ -45,15 +46,15 @@ pub fn apply_rounding_buffer(
         let buy_token_price = match token_prices.get(&buy_token) {
             Some(price) if *price > 0 => *price as f64,
-            _ => 1.0,
+            _ => fee_token_price,
```
Should the default consider the token's decimals? Depending on if a higher price leads to a larger or lesser rounding buffer we should probably choose a price of 10^^(decimals +/- 5 or the like). This would allow to cover prices from 0,0001 to 10000

e00E replied:
Assuming that tokens usually have decimals set so that one base unit is very roughly one USD this would make sense.
I'm thinking a good way to sidestep this here would be to instead of taking the HashMap with prices take a function `TokenId -> Price`, then the caller has to decide how they want to implement it.
In a future PR that has to call this function we can then do something like what you describe.

e00E replied:
Implemented this idea as another commit in this PR. The discussion is still valid but this module won't have to make the decision.

fleupold replied:
We might want to have different default behaviour based on whether it's the buy or sell token. Given the buffer depends on 

`let estimated_xrate = estimated_buy_token_price / estimated_sell_token_price;`

we might want the buy_token_price fallback to something large (10^(decimals+5)) and the sell_token_price fall back to something small (10^(decimals-5))

I think to fine-tune this it would be nice to see what the impact of these changes would be on accuracy. But for now, I think it's fine assuming 10^decimals (we should document somewhere all the places on which we might fine-tune later).

e00E committed:
Force prices to be handled by caller

This makes this part of the code not have to make the decision of what
fallback prices to use.

nlordell commented on price-estimator/src/solver_rounding_buffer.rs:
```diff
@@ -27,34 +25,25 @@ pub fn rounding_buffer(fee_token_price: f64, sell_token_price: f64, buy_token_pr
 #[allow(dead_code)]
 pub fn apply_rounding_buffer(
-    token_prices: &TokenPrices,
+    token_prices: &dyn Fn(TokenId) -> u128,
```
```suggestion
    token_prices: impl Fn(TokenId) -> u128,
```

e00E committed:
Merge branch 'master' into rounding-buffer-no-price
